crates/metal-llvm/src/expr.rs

```diff
@@ -0,0 +1,78 @@
+use std::ffi::{c_uint, CString};
+
+use llvm_sys::{
+    core::{LLVMBuildCall2, LLVMConstInt, LLVMConstStringInContext, LLVMGetNamedFunction},
+    prelude::LLVMValueRef,
+};
+use metal_mir::{
+    expr::{literals::Literal, Expr},
+    parcel::Module,
+    types::primitives::Primitive,
+};
+
+use crate::{CodeGenType, CodeGenValue};
+
+fn args_to_values(llvm: &crate::LLVMRefs, module: &Module, args: &Vec<Expr>) -> Vec<LLVMValueRef> {
+    let mut v = Vec::with_capacity(args.len());
+
+    for arg in args {
+        v.push(arg.codegen_value(llvm, module));
+    }
+
+    v
+}
+
+impl CodeGenValue for Expr {
+    fn codegen_value(
+        &self,
+        llvm: &crate::LLVMRefs,
+        module: &metal_mir::parcel::Module,
+    ) -> llvm_sys::prelude::LLVMValueRef {
+        match self {
+            Self::FunctionCall(fcall) => unsafe {
+                let c_fun_name = CString::new(fcall.name.as_str()).unwrap();
+
+                // TODO: handle possible errors
+                let func = LLVMGetNamedFunction(llvm.module, c_fun_name.as_ptr());
+                assert!(!func.is_null());
+
+                let args_num = fcall.arguments.len();
+                LLVMBuildCall2(
+                    llvm.builder,
+                    fcall.signature.return_type.codegen_type(llvm, module),
+                    func,
+                    args_to_values(llvm, module, &fcall.arguments).as_mut_ptr(),
+                    args_num as c_uint,
+                    c_fun_name.as_ptr(),
+                )
+            },
+            // Has to be cloned since we can't Unpack a referenced Box.
+            // Essentially: `**lit` doesn't work, since Copy isn't implemented
+            // for MIR types.
+            // Context: https://github.com/metallang/metal/pull/51#discussion_r1838137310
+            Self::Literal(lit) => match *lit.clone() {
```

Comment:
This doesn't have to be cloned as an owned value is not required in any of the `match`'s arms.

```suggestion
            Self::Literal(lit) => match &lit {
```